### PR TITLE
User create event

### PR DESCRIPTION
**Pull request by ericnormand**
[Pivotal Card](https://www.pivotaltracker.com/story/show/95886418)

Wire up electorate.create events onto the events exchange. We did it in the entities.electorate namespace because all electorate creation passes through there. Putting it in the handlers is less certain.

Lots of queue wiring. Please please check this (I'm talking to you @tie-rack).

User create event handling needs to look up upcoming elections associated with the users's registered address's ocdids. The elections will have authority-levels. We select the users's ocdids that match the authority levels and search the election-authorities for those ocdids. This gives us the three parts of an electorate (userid, electionid, #{election-authorityid}).

The shakiest part is the function `ocd-id-level` which judges the ocd's level using regexes. :municipal is the fallthrough though we know that not all of those falling through are actually municipal. Suggestions for improvements are welcome.

Several things that could be moved into a separate library:
* channels/start-event-handler! (which starts a thread to listen for events)
* handlers/deref-or-timeout (which derefs promises or throws an exception on timeout)
* handlers/handle-errors (it exists already in other projects but this one catches Throwable, which is the superclass of all errors; other versions only catch runtimeexception, which would let AssertionErrors and other Exceptions through).

### CODE REVIEW

**Comment by ericnormand** [comment]
Assigning to @cap10morgan ; also, please look at it, @tie-rack, because langhor is hard!

File: src/electorate_works/channels.clj [hunk 1]
```diff
@@ -1,8 +1,44 @@
...
+(defonce create-events (async/chan))
+
+;; incoming
+(defonce user-create-events (async/chan))
+
+;; query-response
+(def election-search (async/chan))
+(def election-authority-search (async/chan))
+
 ;;; TODO: Create channels and add them to the list of channels to close.
```
Comment (cap10morgan) [inline src/electorate_works/channels.clj]: We can drop this template TODO comment now, yes?

File: src/electorate_works/entities/electorate.clj [hunk 2]
```diff
@@ -31,7 +39,10 @@
   (s/validate create-schema params)
   (when-let [existing (find-by-ids (db/db) (:user-id params) (:election-id params))]
     (throw (IllegalArgumentException. "Electorate for User and Election already exists")))
-  @(db/transact [(->tx-data params)])
+  (let [{:keys [db-after]} @(db/transact [(->tx-data params)])
+        new-electorate (-> (find-by-ids db-after (:user-id params) (:election-id params))
+                           entity->electorate)]
```
Comment (cap10morgan) [inline src/electorate_works/entities/electorate.clj]: Would this be slightly clearer as below?

```clojure
(-> db-after
     (find-by-ids (:user-id params) (:election-id params))
     entity->electorate)
```

File: src/electorate_works/handlers.clj [hunk 3]
```diff
@@ -53,3 +56,45 @@
...
   {:status :ok})
+
+(def search-elections (k/ch->response-fn channels/election-search))
+(def search-election-authorities (k/ch->response-fn channels/election-authority-search))
+
+(defn ocd-id-level [ocd-id]
+  (cond
+    (re-find #"state:[-\w.~]+$"  ocd-id) :state
+    (re-find #"county:[-\w.~]+$" ocd-id) :county
+    (re-find #"parish:[-\w.~]+$" ocd-id) :county
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: Good thing we have a Louisianan on staff to catch that! ;)

File: src/electorate_works/handlers.clj [hunk 4]
```diff
@@ -53,3 +56,45 @@
...
+      (throw (TimeoutException.
+              (str "Response timed out performing " action-description)))
+      resp)))
+
+(defn create-electorates-from-user
+  [search-elections search-election-authorities user timeout]
+  (let [user-id (:id user)
+        user-ocdids (get-in user [:addresses :registered :ocd-division-id])
+        ocds-by-level (delay (group-by ocd-id-level user-ocdids))
+        mem-search (memoize search-election-authorities)]
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: This is just memoized for *this invocation* of `create-electorates-from-user`? And it's so that we don't waste resources pulling the same authorities across multiple elections for one user?

File: src/electorate_works/handlers.clj [hunk 5]
```diff
@@ -53,3 +56,45 @@
...
+                           (deref-or-timeout timeout "searching elections")
+                           :elections)]
+        (let [{level :authority-level election-id :id} election
+              authorities (-> (mem-search {:district-divisions (get @ocds-by-level level)})
+                              (deref-or-timeout timeout "searching election authorities")
+                              :authorities)
+              authority-ids (set (map :id authorities))]
+          (e/create-electorate {:user-id                user-id
+                                :election-id            election-id
+                                :election-authority-ids authority-ids})))))
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: How can we tell which authority is the election authority and which is the registration authority here?

File: src/electorate_works/queue.clj [hunk 6]
```diff
@@ -38,6 +54,14 @@
       (lq/declare delete-ch
                   "electorate-works.electorate.delete"
                   (config :rabbitmq :queues "electorate-works.delete"))
+
+      (lq/declare election-search-ch
+                  "election-works.election.search"
+                  (config :rabbitmq :queues "election-works.election.search"))
+      (lq/declare election-authority-search-ch
+                  "election-authority-works.search"
+                  (config :rabbitmq :queues "election-authority-works.search"))
```
Comment (cap10morgan) [inline src/electorate_works/queue.clj]: Would it not be more consistent to use `election-authority-works.authority.search` here?

File: src/electorate_works/queue.clj [hunk 7]
```diff
@@ -45,8 +69,19 @@
...
+      (k/wire-up-service election-search-ch "election-works.election.search" channels/election-search)
+      (k/wire-up-service election-authority-search-ch "election-authority-works.search" channels/election-authority-search)
+
       {:connections #{@connection}
-       :channels #{ok-ch create-ch delete-ch}})))
+       :channels #{events-ch
+                   ok-ch create-ch delete-ch
+                   create-events-ch
+                   user-create-ch
+                   election-search-ch election-authority-search-ch}})))
```
Comment (cap10morgan) [inline src/electorate_works/queue.clj]: At some point we (Troy and I and maybe Chris) realized that this probably should be a vector and the order should be "incoming channels first, then outgoing channels" so that they get closed in that order on shutdown. Basically: Turn off all faucets before you plug the drains.

Yes, this needs to go into docs and templates soon. :)

File: src/electorate_works/queue.clj (hunk 6 above)
Comment (tank157) [inline src/electorate_works/queue.clj]: Also, I noted this because I'm working on the `election.create`, but the config has neither `election-authority-works.search` nor `election-works.election.search` in it currently.

File: test/electorate_works/handlers_test.clj [hunk 8]
```diff
@@ -12,9 +14,9 @@
 
 (use-fixtures :each
   (fn [f]
-     (d/delete-database (config :datomic :uri))
-     (db/initialize)
-     (f)))
+    (d/delete-database (config :datomic :uri))
```
Comment (cap10morgan) [inline test/electorate_works/handlers_test.clj]: You didn't *really* touch this, but you kind of did. So:

Cognitect told us that deleting and re-creating the same Datomic DB is an anti-pattern. It flies against Datomic's rejection of "naming a place where you keep updating the truth" vs. having a time-based log of unchanging facts. They recommend either embracing time in your test database (e.g. reverting all facts on cleanup) or creating a DB with a unique name each time if you really want to guarantee an empty DB for all *t*. Otherwise we run into the problem we were seeing before w/ ballot-scout (IIRC) where this was exhibiting race-condition-like behavior when running tests in docker-compose.

File: test/electorate_works/handlers_test.clj [hunk 9]
```diff
@@ -91,3 +115,55 @@
...
           (let [delete-response (handler good-message)]
             (is (= :ok (:status delete-response)))))))))
+
+(deftest ocd-id-level-test
+  (testing "should be :state"
+    (is (= :state (ocd-id-level "ocd-division/country:us/state:ar"))))
+  (testing "should be :county"
+    (is (= :county (ocd-id-level "ocd-division/country:us/state:ar/county:ashley"))))
+  (testing "council_district is municipal?"
+    (is (= :municipal (ocd-id-level "ocd-division/country:us/state:ar/county:ashley/council_district:2")))))
```
Comment (cap10morgan) [inline test/electorate_works/handlers_test.clj]: Good question. We should plug some addresses into Google's API and start getting a feel for what kinds of values it returns.

But, IIRC, a city council district ID would look more like: `ocd-division/country:us/state:ar/place:little-rock/council_district:2`

But if there are county-level legislative bodies (hmm...), then I suppose you could have districts there too.

I think that we don't need to worry about that, however, because the authority-levels will be manually set on the elections (so even if an election is only for a handful of state leg. districts, for example, the levels will be set to `:county` or whatever is appropriate there). And then the users will all have an OCD-ID that ends in their county among the set returned by Google, and that will match up to the appropriate authorities.

In other words, I don't think we need to group all OCD-IDs into levels, and we encounter some surprising / incorrect behavior if we do. We may want to start by just mapping the following:

OCD-IDs that end in `/state:__` -> `:state`
                               `/county:__` or `/parish:__` -> `:county`
                               `/place:__` -> `:municipal`

...and leave the rest w/o levels.

I'd be happy to talk through this if it doesn't make sense.

**Comment by cap10morgan** [comment]
Here are the OCD-IDs returned by Google for my home address:

```
ocd-division/country:us/state:co/sldu:34"
ocd-division/country:us/state:co/cd:1
ocd-division/country:us/state:co/county:denver/council_district:9
ocd-division/country:us/state:co
ocd-division/country:us
ocd-division/country:us/state:co/county:denver
  (this also returns an "also-known-as" sub-attribute with: "ocd-division/country:us/state:co/place:denver" because Denver is both a city and a county)
ocd-division/country:us/state:co/sldl:5
```

You can try this out for yourself here: https://developers.google.com/civic-information/docs/v2/representatives/representativeInfoByAddress#try-it

File: src/electorate_works/handlers.clj (hunk 4 above)
Comment (ericnormand) [inline src/electorate_works/handlers.clj]: Yes.

Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: :+1:

File: test/electorate_works/handlers_test.clj (hunk 8 above)
Comment (ericnormand) [inline test/electorate_works/handlers_test.clj]: I'll make a card because this is a sweeping change across all of our DBs. I won't address it here.

I actually think it's the right way to do it. We generate databases with unique names (in memory) for each test. It will be hard because we're using the global config data.

File: test/electorate_works/handlers_test.clj (hunk 9 above)
Comment (ericnormand) [inline test/electorate_works/handlers_test.clj]: I was actually thinking about this. Get ready for a patch.

File: test/electorate_works/handlers_test.clj (hunk 8 above)
Comment (cap10morgan) [inline test/electorate_works/handlers_test.clj]: Sorry, I'm not clear on what you think is the right way to do it. Are you saying we *should* generate databases with unique names in memory for each test? If so, that seems reasonable to me.

But yes, we'll have to override the config (something I've wanted w/ resource-config elsewhere) and it should be done across the board. A new card sounds good.

File: src/electorate_works/queue.clj (hunk 6 above)
Comment (ericnormand) [inline src/electorate_works/queue.clj]: @cap10morgan Yes, but that is the name of the queue in the other repo.

Comment (ericnormand) [inline src/electorate_works/queue.clj]: @tank157 Fixed.

File: src/electorate_works/queue.clj (hunk 7 above)
Comment (ericnormand) [inline src/electorate_works/queue.clj]: Ok, done!

File: test/electorate_works/handlers_test.clj (hunk 8 above)
Comment (ericnormand) [inline test/electorate_works/handlers_test.clj]: Creating new in-memory databases in an :each fixture is **The Right Way<sup>:tm:</sup>** to do it.

Comment (ericnormand) [inline test/electorate_works/handlers_test.clj]: I'm not sure if overriding config is the right level of abstraction. Seems like we would want to override datomic-toolbox's choice of database.

**Comment by ericnormand** [comment]
Ready for a new review. I think it's a pretty big change again. It shouldn't be rushed. I will understand if it waits till next sprint for approval.

File: src/electorate_works/queue.clj (hunk 6 above)
Comment (tank157) [inline src/electorate_works/queue.clj]: Cool. Coolcoolcool.

**Comment by tank157** [comment]
Once [this change](https://github.com/democracyworks/user-works/pull/9) over in user-works is merged, we'll need to revisit this and change `:ocd-division-id` to `:ocd-divisions`, but that hasn't happened yet, and should probably wait for this and the PR for `election.create` to get merged first anyway.

**Comment by ericnormand** [comment]
What else is involved besides vote-by-mail?

**Comment by cap10morgan** [comment]
@ericnormand 

Right now in our code: Maybe nothing.

But conceptually: Literally everything else about running the election. Polling places, ballots, dates, deadlines, etc.

I just don't want to name something after a subset of what it is.

**Comment by ericnormand** [comment]
Makes sense.

**Comment by cap10morgan** [comment]
LGTM
